expr/func_aspercent.go

```diff
@@ -0,0 +1,258 @@
+package expr
+
+import (
+	"errors"
+	"fmt"
+	"math"
+	"sort"
+	"strings"
+
+	"github.com/grafana/metrictank/api/models"
+	schema "gopkg.in/raintank/schema.v1"
+)
+
+type FuncAsPercent struct {
+	in          GraphiteFunc
+	totalFloat  float64
+	totalSeries GraphiteFunc
+	nodes       []expr
+}
+
+func NewAsPercent() GraphiteFunc {
+	return &FuncAsPercent{totalFloat: math.NaN()}
+}
+
+func (s *FuncAsPercent) Signature() ([]Arg, []Arg) {
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgIn{
+			key: "total",
+			opt: true,
+			args: []Arg{
+				ArgFloat{val: &s.totalFloat},
+				ArgSeriesList{val: &s.totalSeries},
+			},
+		},
+		ArgStringsOrInts{val: &s.nodes, opt: true, key: "nodes"},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncAsPercent) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncAsPercent) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	var outSeries []models.Series
+	defer func() { cache[Req{}] = append(cache[Req{}], outSeries...) }()
+	var totals []models.Series
+	if s.totalSeries != nil {
+		totals, err = s.totalSeries.Exec(cache)
+		if err != nil {
+			return nil, err
+		}
+		if len(totals) == 0 {
+			totals = nil
+		}
+	}
+
+	if s.nodes != nil {
+		// Set of keys
+		keys := make(map[string]struct{})
+		// Series grouped by key
+		metaSeries := groupSeriesByKey(series, s.nodes, &keys)
+		// The totals series for each key
+		var totalSeries map[string]models.Series
+
+		// calculate the sum
+		if math.IsNaN(s.totalFloat) && totals == nil {
+			totalSeries = getTotalSeries(metaSeries)
+			// calculate sum of totals series
+		} else if totals != nil {
+			totalSeriesLists := groupSeriesByKey(totals, s.nodes, &keys)
+			totalSeries = getTotalSeries(totalSeriesLists)
+		} else {
+			return nil, errors.New("total must be None or a seriesList")
```

Comment:
seems like it's fairly easy to trigger this by specifying a serieslist pattern that doesn't match any series.
maybe in that case we can provide a better error message? (basically above at `if len(totals) == 0 {` maybe directly return an error ? what does graphite do in this case?

Reply:
I'm not sure what you mean here. This error message gets triggered if they pass in a number. In the case where they pass in a nodes argument, only a series or nothing should be passed in for the `total` argument. First branch is triggered if neither is passed in. Second branch is triggered if a series is passed in. Which leaves the case where a number is passed in (i.e `math.IsNaN(s.totalFloat) = false`)

Graphite throws that exact error message is that case.

Reply:
if you specify a serieslist pattern that doesn't match any series, then I believe this will happen:
```
        if s.totalSeries != nil {
                totals, err = s.totalSeries.Exec(cache)
                if err != nil {
                        return nil, err
                }
                if len(totals) == 0 {
                        totals = nil <---- this right here
                }
        }
```

Reply:
FWIW, Graphite crashes if you do that.

Let me see if I can handle it in some clean way

Reply:
Ok, so this was just unnecessary:
```
if len(totals) == 0 {
            totals = nil
}
```

So I removed it. Now if a series returns empty, it does not assume that there were no arguments